### PR TITLE
DEV | AuthService Refactor

### PR DESCRIPTION
**Pull request by bseptember**
### Overview

Added unit test for AuthService.cs
Refactored AuthService.cs according to [Azure board ticket](https://dev.azure.com/retro-rabbit/HRIS/_sprints/taskboard/HRIS%20Team/HRIS/Sprint%2025?System.AssignedTo=bseptember%40retrorabbit.co.za&workitem=9437)

image

### CODE REVIEW

**Review by Wimmas — APPROVED** [review]

File: HRIS.Services.Tests/Services/AuthServiceUnitTest.cs [hunk 1]
```diff
@@ -1,269 +1,963 @@
...
+    {
+        var field = instance.GetType().GetField(fieldName, BindingFlags.NonPublic | BindingFlags.Instance);
+        field?.SetValue(instance, value);
+    }
+
+    private string CreateJwtToken(bool notExpired = true, bool hasExpClaim = true)
+    {
+        var securityKey = new SymmetricSecurityKey(System.Text.Encoding.Default.GetBytes("test_secret_key_which_must_be_a_longer_than_128bits"));
+
+        // Note for developer: We use an asymmetric key, but for testing we can simply use a symmetric key!
```
Comment (madwizzard) [inline HRIS.Services.Tests/Services/AuthServiceUnitTest.cs]:
```suggestion
```

File: HRIS.Services/Services/AuthService.cs [hunk 2]
```diff
@@ -6,44 +6,46 @@
...
-    private static string _issuer = string.Empty;
-    private static string _audience = string.Empty;
-
-    public AuthService(IErrorLoggingService errorLoggingService, IOptions<AuthManagement> options)
+    private string _cachedAccessToken;
+    private string _clientId;
+    private string _clientSecret;
+    private string _issuer;
+    private string _audience;
+    private HttpClient _httpClient;
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
    private readonly string _clientId;
    private readonly string _clientSecret;
    private readonly string _issuer;
    private readonly string _audience;
    private readonly HttpClient _httpClient;
```

File: HRIS.Services/Services/AuthService.cs [hunk 3]
```diff
@@ -6,44 +6,46 @@
...
 using Microsoft.Extensions.Options;
 using System.IdentityModel.Tokens.Jwt;
+using System.Text.Json;
 
 namespace HRIS.Services.Services;
 
 public class AuthService : IAuthService
 {
-    private readonly IErrorLoggingService _errorLoggingService;
+    private readonly IManagementApiClient _managementApiClient;
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs:15]:
```suggestion
```

File: HRIS.Services/Services/AuthService.cs [hunk 4]
```diff
@@ -6,44 +6,46 @@
...
     {
-        _errorLoggingService = errorLoggingService;
         _authManagement = options.Value;
         _clientId = _authManagement.ClientId ?? EnvironmentVariableHelper.AUTH_MANAGEMENT_CLIENT_ID;
         _clientSecret = _authManagement.ClientSecret ?? EnvironmentVariableHelper.AUTH_MANAGEMENT_CLIENT_SECRET;
         _issuer = _authManagement.Issuer ?? EnvironmentVariableHelper.AUTH_MANAGEMENT_ISSUER;
         _audience = _authManagement.Audience ?? EnvironmentVariableHelper.AUTH_MANAGEMENT_AUDIENCE;
+        _cachedAccessToken = string.Empty;
+        _managementApiClient = new ManagementApiClient(_cachedAccessToken, new Uri($"{_issuer}api/v2"));
+        _httpClient = new HttpClient();
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
        var authManagement = options.Value;
        _clientId = authManagement.ClientId ?? EnvironmentVariableHelper.AUTH_MANAGEMENT_CLIENT_ID;
        _clientSecret = authManagement.ClientSecret ?? EnvironmentVariableHelper.AUTH_MANAGEMENT_CLIENT_SECRET;
        _issuer = authManagement.Issuer ?? EnvironmentVariableHelper.AUTH_MANAGEMENT_ISSUER;
        _audience = authManagement.Audience ?? EnvironmentVariableHelper.AUTH_MANAGEMENT_AUDIENCE;
        _cachedAccessToken = string.Empty;
        _managementApiClient = new ManagementApiClient(_cachedAccessToken, new Uri($"{_issuer}api/v2"));
        _httpClient = new HttpClient();
```

Comment (EduanO) [reply HRIS.Services/Services/AuthService.cs]: Can be a local variable in constructor

File: HRIS.Services/Services/AuthService.cs [hunk 5]
```diff
@@ -65,173 +67,145 @@ private static bool IsTokenExpired(string token)
...
-
-        var response = await client.PostAsync(tokenEndpoint, new FormUrlEncodedContent(requestBody));
+        var response = await _httpClient.PostAsync(tokenEndpoint, new FormUrlEncodedContent(requestBody));
         var content = await response.Content.ReadAsStringAsync();
 
-        if (response.IsSuccessStatusCode)
+        if (!response.IsSuccessStatusCode)
+            throw new CustomException($"Failed response from auth provider: {content}");
+
+        var jsonDoc = System.Text.Json.JsonDocument.Parse(content);
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
        var jsonDoc = JsonDocument.Parse(content);
```

File: HRIS.Services/Services/AuthService.cs [hunk 6]
```diff
@@ -65,173 +67,145 @@ private static bool IsTokenExpired(string token)
...
-        else
-        {
-            throw new Exception($"Failed response from auth provider: {content}");
-        }
+
+        throw new CustomException($"Failed response from auth provider. access_token key not found.");
     }
 
-    public async Task<IPagedList<Role>> GetAllRolesAsync()
+    public async Task<IPagedList<Auth0.ManagementApi.Models.Role>> GetAllRolesAsync()
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
    public async Task<IPagedList<Role>> GetAllRolesAsync()
```

File: HRIS.Services/Services/AuthService.cs [hunk 7]
```diff
@@ -65,173 +67,145 @@ private static bool IsTokenExpired(string token)
...
+
+        var allRoles = await _managementApiClient.Roles.GetAllAsync(request, pagination);
+        if (allRoles == null)
+            throw new CustomException("Failed to retrieve roles.");
+
+        return allRoles;
     }
 
-    public async Task<IPagedList<User>> GetAllUsersAsync()
+    public async Task<IPagedList<Auth0.ManagementApi.Models.User>> GetAllUsersAsync()
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
    public async Task<IPagedList<User>> GetAllUsersAsync()
```

File: HRIS.Services/Services/AuthService.cs [hunk 8]
```diff
@@ -65,173 +67,145 @@ private static bool IsTokenExpired(string token)
...
+
+        var allUsers = await _managementApiClient.Users.GetAllAsync(request, pagination);
+        if (allUsers == null)
+            throw new CustomException("Failed to retrieve roles.");
+
+        return allUsers;
     }
 
-    public async Task<IList<User>> GetUsersByEmailAsync(string email)
+    public async Task<IList<Auth0.ManagementApi.Models.User>> GetUsersByEmailAsync(string email)
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
    public async Task<IList<User>> GetUsersByEmailAsync(string email)
```

File: HRIS.Services/Services/AuthService.cs [hunk 9]
```diff
@@ -65,173 +67,145 @@ private static bool IsTokenExpired(string token)
...
+
+        var user = await _managementApiClient.Users.GetUsersByEmailAsync(email);
+        if (user == null)
+            throw new CustomException("Failed to retrieve user by email.");
+
+        return user;
     }
 
-    public async Task<IPagedList<AssignedUser>> GetUsersByRoleAsync(string roleId)
+    public async Task<IPagedList<Auth0.ManagementApi.Models.AssignedUser>> GetUsersByRoleAsync(string roleId)
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
    public async Task<IPagedList<AssignedUser>> GetUsersByRoleAsync(string roleId)
```

File: HRIS.Services/Services/AuthService.cs [hunk 10]
```diff
@@ -65,173 +67,145 @@ private static bool IsTokenExpired(string token)
...
+
+        var users = await _managementApiClient.Roles.GetUsersAsync(roleId, new PaginationInfo(0, 50));
+        if (users == null)
+            throw new CustomException("Failed to retrieve users.");
+
+        return users;
     }
 
-    public async Task<IPagedList<Role>> GetUserRolesAsync(string userId)
+    public async Task<IPagedList<Auth0.ManagementApi.Models.Role>> GetUserRolesAsync(string userId)
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
    public async Task<IPagedList<Role>> GetUserRolesAsync(string userId)
```

File: HRIS.Services/Services/AuthService.cs [hunk 11]
```diff
@@ -248,244 +222,169 @@ public async Task<Role> CreateRoleAsync(string roleName, string description)
...
 
         bool roleFound = allRoles.Any(role => role.Id == roleId);
-        if (roleFound)
+        if (!roleFound)
+            return false;
+
+        var token = await GetAuth0ManagementAccessToken();
+        _managementApiClient.UpdateAccessToken(token);
+
+        var temporaryRole = await _managementApiClient.Roles.UpdateAsync(roleId, new RoleUpdateRequest
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
        await _managementApiClient.Roles.UpdateAsync(roleId, new RoleUpdateRequest
```

File: HRIS.Services/Services/AuthService.cs [hunk 12]
```diff
@@ -248,244 +222,169 @@ public async Task<Role> CreateRoleAsync(string roleName, string description)
...
-        if (existingUser != null)
-        {
-            await client.Users.DeleteAsync(userId);
-        }
+        _managementApiClient.UpdateAccessToken(token);
+
+        if (string.IsNullOrEmpty(userId))
+            return false;
+
+        var existingUser = await _managementApiClient.Users.GetAsync(userId, null, true);
```
Comment (EduanO) [inline HRIS.Services/Services/AuthService.cs]:
```suggestion
        var existingUser = await _managementApiClient.Users.GetAsync(userId);
```

Comment (EduanO) [reply HRIS.Services/Services/AuthService.cs]: They are already set to default `null` and `true`

**Review by EduanO — CHANGES_REQUESTED** [review]

**Review by nzikalala — APPROVED** [review]

File: HRIS.Services/Services/AuthService.cs (hunk 3 above)
Comment (bseptember) [reply HRIS.Services/Services/AuthService.cs:15]: @EduanO  is this a mistake? _managementApiClient is required in the code

Comment (EduanO) [reply HRIS.Services/Services/AuthService.cs:15]: Yeah not sure where I was going with that. My bad.

**Review by EduanO — APPROVED** [review]

**Pull request merged**